test: Support decoding segwit address in address_to_scriptpubkey()

ismaelsadeeq opened the pull request:
[rpc_scantxoutset.py](https://github.com/bitcoin/bitcoin/blob/e695d8536e534f1e59de4a62a0d87a93e7a73456/test/functional/rpc_scantxoutset.py#L26)  sendtodestination only sends to legacy addresses and scriptPubkeys because  [wallet.py](https://github.com/bitcoin/bitcoin/blob/e695d8536e534f1e59de4a62a0d87a93e7a73456/test/functional/test_framework/wallet.py#L415) address_to_scriptpubkey does not support conversion of segwit address.

This update enables address_to_scriptpubkey to support the conversion of testnet segwit addresses to scriptPubkeys.

This change will enable [rpc_scantxoutset.py](https://github.com/bitcoin/bitcoin/blob/e695d8536e534f1e59de4a62a0d87a93e7a73456/test/functional/rpc_scantxoutset.py#L22) ScantxoutsetTest to have more test coverage by adding more sendtodestination calls with bech32 and bech32m testnet addresses, then test the bech32 and bech32m  derivation subsets UTXO amount in [Test extended key derivation](https://github.com/bitcoin/bitcoin/blob/e695d8536e534f1e59de4a62a0d87a93e7a73456/test/functional/rpc_scantxoutset.py#L84).

I will add the test coverage in a subsequent Pull request.

willcl-ark commented:
Concept ACK

Thanks for working on this. Increasing the address to scriptpubkey decoding tests to include bech32 seems desirable. I think this would mean that we could remove the `todo` on [L427](https://github.com/bitcoin/bitcoin/pull/27269/commits/4b95be0c496947606aaa661507ca06d53cc358c5#diff-7932a60a9127fd22d10d367526fd7b987f9647ce017595f8b1af5c32d5db0083R427) of wallet.py too?

The way the equivalent base58 function `test_base58encodedecode()` is implemented is slightly different from the new `test_bech32_decode()`:

1. It checks two functions simultaneously, `base58_to_bytes()` and `byte_to_base58()`, whereas your first commit only adds `bech32_to_bytes()` and not the equivalent `byte_to_bech32()`.
2. The test vectors provided to the new test are raw address strings (and raw bytes) vs hex-encoded bytestrings in the base58 version.

I think for completion it could be worth adding the equivalent `byte_to_bech32()` function and matching the test vector format used in the base58 checks, seeing as you are here making this change now.

Woudl be curious what others think on this though...

ismaelsadeeq commented:
Yes, I agree the todo can be removed since there are no other address types to add.

- I think there is no need to add `byte_to_bech32()`, there is an existing encode_segwit_address() function that does the same thing.
- I can use  bech32_to_bytes()  and encode_segwit_address() with hex encoded string and version as test vectors in  test_bech32_decode() just like in test_base58encodedecode().

ismaelsadeeq commented:
Thanks for reviewing, I made the update.

josibake commented on test/functional/test_framework/address.py:
```diff
@@ -158,6 +158,11 @@ def check_script(script):
         return script
     assert False
 
+def bech32_to_bytes(address):
```
styling nit: two blank lines between functions, per PEP8 guidelines

ismaelsadeeq replied:
I think for consistency in the code base it's okay.
There is only one blank line between functions in address.py

josibake replied:
The goal is to have the code base consistent with our style guide, not consistent with itself. This means any new code should adhere to the style guide and, if possible, old code should be formatted consistent with the style guide if it's already being touched as part of a PR.

josibake commented on test/functional/test_framework/address.py:
```diff
@@ -20,8 +20,8 @@
 )
-from .segwit_addr import encode_segwit_address
 from .util import assert_equal
+from test_framework.segwit_addr import (encode_segwit_address, bech32_decode, convertbits)
```
also, if you end up taking my suggestion to use `decode_segwit_address`, you can remove the `bech32_decode, convertbits` imports

josibake commented on test/functional/test_framework/wallet.py:
```diff
@@ -50,6 +51,7 @@
     key_to_p2wpkh_script,
     keyhash_to_p2pkh_script,
     scripthash_to_p2sh_script,
+    program_to_witness_script
```
styling nit: should be lexicographically sorted and add a comma at the end of each line (even if it is the last one)

ismaelsadeeq replied:
Thanks. :+1:

josibake reviewed (commented):
Concept ACK

Thanks for tackling a TODO! I'm guessing there are other places in the functional tests where we would want coverage for more than just base58 addresses, so definitely a worthwhile change. Overall, looks good! I left a few suggestions and style comments in my review. Mainly:

* I think for consistency you can use the `decode_segwit_address` function from `segwit_addr.py`. This avoids having the logic for decoding bech32 addresses in multiple places.
* Even though a lot of these files don't adhere to PEP8 guidelines, we want to make sure any new code introduced does follow the guidelines. What I typically do is run `flake8` on the file I plan to work on before making any changes and then run `flake8` again after my changes to make sure all of my new changes follow PEP8 guidelines

ismaelsadeeq commented on test/functional/test_framework/address.py:
```diff
@@ -158,6 +158,11 @@ def check_script(script):
     assert False
 
+def bech32_to_bytes(address):
+    encoding, hrp, data = bech32_decode(address)
+    payload = bytes(convertbits(data[1:], 5, 8, False))
+    version = data[0]
+    return version, payload
```
Thanks for pointing that out :100:

ismaelsadeeq commented on test/functional/test_framework/address.py:
```diff
@@ -176,3 +181,18 @@ def check_base58(data, version):
         check_base58(bytes.fromhex('0041c1eaf111802559bad61b60d62b1f897c63928a'), 0)
         check_base58(bytes.fromhex('000041c1eaf111802559bad61b60d62b1f897c63928a'), 0)
         check_base58(bytes.fromhex('00000041c1eaf111802559bad61b60d62b1f897c63928a'), 0)
+
```
functions in the codebase have one blank line in between.

josibake replied:
This is because a lot of code was written before we introduced a style guide, or because it was missed during review. If something is not covered by the style guide, then we should try to be consistent with the surrounding code. If something _is_ covered by the style guide, we should be consistent with the style guide, even if that is not consistent with the surrounding code. This helps nudge the codebase towards the style guide over time.

ismaelsadeeq replied:
Thanks

willcl-ark replied:
I may be mistaken but I believe that actually this [_should_](https://peps.python.org/pep-0008/#blank-lines) have 1 line gap, as it's a member function? But its OK, I wouldn't change and invalidate the ACKS you've got already! (I configured my editor to show PEP8 diagnostics so I can view them directly as I'm editing).

josibake commented on test/functional/test_framework/address.py:
```diff
@@ -20,8 +20,11 @@
-from .segwit_addr import encode_segwit_address
 from .util import assert_equal
+from test_framework.segwit_addr import (
+    encode_segwit_address,
```
style nit: these should be lexicographically sorted

ismaelsadeeq replied:
Okay :+1:

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/9aac917c6e7fae48ceae5fe786c412a583e46d3e

Thanks for incorporating the feedback and addressing the style comments! Looks good.

josibake commented on test/functional/test_framework/wallet.py:
```diff
@@ -414,11 +416,13 @@ def getnewdestination(address_type='bech32m'):
 
 def address_to_scriptpubkey(address):
     """Converts a given address to the corresponding output script (scriptPubKey)."""
+    if address[:4] == 'tb1q': # testnet segwit
```
thinking about this a bit more, I think we should have a more robust way of checking this. iirc, taproot would be `tb1p` on testnet, so this would not work for segwit v1 addresses.

Maybe instead you could do something like `version, payload = bech32_to_bytes` and then check the version. If the version is None, then move on to `payload, version = base58_to_byte` etc.

ismaelsadeeq replied:
Thanks for the thorough review Josi.
I will fix that.
This means I have to check in `bech32_to_bytes `also for a None version value after   so as not to call bytearray(payload) with None.

josibake commented on test/functional/test_framework/address.py:
```diff
@@ -159,6 +162,14 @@ def check_script(script):
 
+def bech32_to_bytes(address):
+    hrp = address[:2]
+    version, payload = decode_segwit_address(hrp,address)
+    if version is None:
+        return None, None
```
there should be a space between function arguments and I'd return a tuple here to stay consistent with the other functions in the `test_framework/segwit_addr.py` file:

```suggestion
    version, payload = decode_segwit_address(hrp, address)
    if version is None:
        return (None, None)
```

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/9e377f481bd0496d868ea0b21f53d2c9bd0ef85a

nice work!

theStack reviewed (commented):
Concept ACK

Thanks for tackling this, and warm welcome as a new contributor!
With this change, we don't have to issue `getaddressinfo` or `validateaddress` RPCs anymore to convert from address to scriptPubKey.

To ensure that the helper `address_to_scriptpubkey` works as expected, I think it would be good to call it least once with a segwit address in a functional test. One possible candidate is e.g. replacing the RPC call in this line:
https://github.com/bitcoin/bitcoin/blob/4c6b7d330a4e80460dcce19b1a0a47d77a0b99ea/test/functional/wallet_fast_rescan.py#L61
This reveals that the added decoding support currently doesn't work as expected, see review comment below.

ismaelsadeeq commented on test/functional/test_framework/address.py:
```diff
@@ -159,6 +162,14 @@ def check_script(script):
 
 
+def bech32_to_bytes(address):
+    hrp = address[:2]
```
Thank you for your valuable feedback.
- I will use `.split('1')` on the address to extract the hrp, and also assert to verify if the hrp is valid as you noted.
- I will have to modify `address_to_scriptpubkey ` to check for base58 address first before calling `bech32_to_bytes` or it will assert false.

I hope this addresses the issue

josibake replied:
I'm not sure I understand your second point, @ismaelsadeeq .  The order shouldn't matter in `address_to_scriptpubkey`. It should be sufficient to properly extract the HRP and check that it is a bech32.

ismaelsadeeq replied:
`bech32_to_bytes`  will be updated to handle all segwit addresses and check if hrp is valid.
```python
def bech32_to_bytes(address):
    hrp = address.split('1')[0]
    assert hrp in ['bc', 'tb', 'bcrt'], f"hrp should be 'bc', 'tb', or 'bcrt', not '{hrp}'"
    version, payload = decode_segwit_address(hrp, address)
    if version is None:
        return (None, None)
    return version, bytearray(payload)
```
So calling address_to_scriptpubkey with a base58 address will assert to false, because the hrp is invalid.
```python
def address_to_scriptpubkey(address):
    """Converts a given address to the corresponding output script (scriptPubKey)."""
    version, payload = bech32_to_bytes(address)
    if version is not None:
        return program_to_witness_script(version, payload) # testnet segwit scriptpubkey
    payload, version = base58_to_byte(address)
    if version == 111:  # testnet pubkey hash
        return keyhash_to_p2pkh_script(payload)
    elif version == 196:  # testnet script hash
        return scripthash_to_p2sh_script(payload)
    # TODO: also support other address formats
    else:
        assert False
``` 
`address_to_scriptpubkey` will be updated to check whether an address is in base58 before calling `bech32_to_bytes` 
```python
def address_to_scriptpubkey(address):
    """Converts a given address to the corresponding output script (scriptPubKey)."""
    prefix = address[0]
    if prefix in ['m','n','2']:
        payload, version = base58_to_byte(address)
        if version == 111:  # testnet pubkey hash
            return keyhash_to_p2pkh_script(payload)
        elif version == 196:  # testnet script hash
            return scripthash_to_p2sh_script(payload)

    version, payload = bech32_to_bytes(address)
    return program_to_witness_script(version, payload) #segwit scriptpubkey
```

ismaelsadeeq replied:
That will work also, thanks @josibake

theStack replied:
@josibake: yes, agree that considering on what we need in `address_to_scriptpubkey` returning `(None, None)` is better than asserting. LGTM!

ismaelsadeeq commented on test/functional/test_framework/wallet.py:
```diff
@@ -414,11 +416,13 @@ def getnewdestination(address_type='bech32m'):
 
 def address_to_scriptpubkey(address):
     """Converts a given address to the corresponding output script (scriptPubKey)."""
+    version, payload = bech32_to_bytes(address)
```
```suggestion
    version, payload = bech32_to_bytes(address)
```
@josibake  this will call will assert to false if the address is a base58 address.

ismaelsadeeq committed:
test: test_bech32_decode in address.py

Adds bech32_to_bytes() which can decode a bech32 address and return the
version as an `int` and the payload in bytes.

bech32_to_bytes() is used by the test_bech32_decode unit test to test
decoding of segwit addresses.

ismaelsadeeq committed:
test: add bech32 decoding support to address_to_scriptpubkey()

This permits functional tests to decode bech32 addresses to scriptpubkeys.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/27269/commits/d178082996dc3000f42816f89afcf3fa4d31e159

I tested this by making sure the `unittests` run and also modifying `test/functional/wallet_fast_rescan.py` to use `address_to_scriptpubkey`, per @theStack 's suggestion. Works great!

I'd love to see a follow-up PR targeting areas where we are a) only testing base58 addresses but should be testing bech32 as well and b) anywhere we are making a `getaddressinfo` RPC call to get the spk to instead use `address_to_scriptpubkey`

theStack reviewed (approved):
ACK d178082996dc3000f42816f89afcf3fa4d31e159 ✔️ 

> I'd love to see a follow-up PR targeting areas where we are a) only testing base58 addresses but should be testing bech32 as well and b) anywhere we are making a getaddressinfo RPC call to get the spk to instead use address_to_scriptpubkey

👍  If anyone wants to work on this, feel free to tag me as reviewer.

josibake commented:
review-beg: @willcl-ark

willcl-ark commented on test/functional/test_framework/address.py:
```diff
@@ -159,6 +162,16 @@ def check_script(script):
 
+def bech32_to_bytes(address):
+    hrp = address.split('1')[0]
+    if hrp not in ['bc', 'tb', 'bcrt']:
```
I think this is totally fine as is here for now, but in the future we may want to put these all in a list/enum

willcl-ark commented on test/functional/test_framework/address.py:
```diff
@@ -176,3 +189,18 @@ def check_base58(data, version):
+
+    def test_bech32_decode(self):
+        def check_bech32_decode(payload, version):
+            hrp = "tb"
```
Instead of hardcoding the hrp here, why not have `check_bech32_decode()` take `hrp` as a parameter, then we can add `bc1` test vectors instead of being limited to `tb1`?

willcl-ark reviewed (approved):
ACK d17808299

I also tested two of the BIP173 [Test Vectors](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#test-vectors):

```python
check_bech32_decode(bytes.fromhex('0020000000c4a5cad46221b2a187905e5266362b99d5e91c6ce24d165dab93e86433'), 1)
check_bech32_decode(bytes.fromhex('00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262'), 1)
```

...which also passed. In doing so I noticed the hardcoding of human readable part `hrp` in _address.py_#L196 was limiting which test vectors I could use, and I think un-hardcoding this  would be great to do as a followup.
